sdks/go/pkg/beam/example_schema_test.go

```diff
@@ -0,0 +1,254 @@
+				if i == 1 {
+					return v.d == nil
+				}
+				return false
+			}, w); err != nil {
+				return err
+			}
+			// Afterwards, each field is encoded using the appropriate helper.
+			if err := coder.EncodeVarUint64(v.c, w); err != nil {
+				return err
+			}
+			// Nil fields have nothing written for them other than the header.
+			if v.d != nil {
+				if err := coder.EncodeDouble(float64(*v.d), w); err != nil {
+					return err
+				}
+			}
+			return nil
+		}, nil
+	case typeΕλληνικά:
+		return func(iface interface{}, w io.Writer) error {
+			// Since the representation for Ελληνικά never has nil fields
+			// we can use the simple header helper.
+			if err := coder.WriteSimpleRowHeader(1, w); err != nil {
+				return err
+			}
+			v := iface.(*Ελληνικά)
+			if err := coder.EncodeStringUTF8(v.q, w); err != nil {
+				return fmt.Errorf("decoding string field A: %v", err)
+			}
+			return nil
+		}, nil
+	}
+	return nil, fmt.Errorf("Unknown Alphabet: %v", rt)
+}
+
+// BuildDecoder returns beam schema decoder functions for types with the Alphabet interface.
+func (p *AlphabetProvider) BuildDecoder(rt reflect.Type) (func(io.Reader) (interface{}, error), error) {
+	switch rt {
+	case typeCyrillic:
+		if p.dec == nil {
+			p.dec = &coder.RowDecoderBuilder{}
+		}
+		// Since Cyrillic is by default encodable, defer to the standard schema row decoder for the type.
+		return p.dec.Build(rt)
+	case typeLatin:
+		return func(r io.Reader) (interface{}, error) {
+			// Since the d field can be nil, we use the header get the nil bits.
+			n, nils, err := coder.ReadRowHeader(r)
+			if err != nil {
+				return nil, err
+			}
+			// Header returns the number of fields, so we check if it has what we
+			// expect. This allows schemas to evolve if necessary.
+			if n != 2 {
+				return nil, fmt.Errorf("expected 2 fields, but got %v", n)
+			}
+			c, err := coder.DecodeVarUint64(r)
+			if err != nil {
+				return nil, err
+			}
+			// Check if the field is nil before trying to decode a value for it.
+			var d *float32
+			if !coder.IsFieldNil(nils, 1) {
+				f, err := coder.DecodeDouble(r)
+				if err != nil {
+					return nil, err
+				}
+				f32 := float32(f)
+				d = &f32
+			}
+			return &Latin{
+				c: c,
+				d: d,
+			}, nil
+		}, nil
+	case typeΕλληνικά:
+		return func(r io.Reader) (interface{}, error) {
+			// Since the representation for Ελληνικά never has nil fields
+			// we can use the simple header helper. Returns an error if
+			// something unexpected occurs.
+			if err := coder.ReadSimpleRowHeader(1, r); err != nil {
+				return nil, err
+			}
+			q, err := coder.DecodeStringUTF8(r)
+			if err != nil {
+				return nil, fmt.Errorf("decoding string field A: %v", err)
+			}
+			return &Ελληνικά{
+				q: q,
+			}, nil
+		}, nil
+	}
+	return nil, nil
+}
+
+// Schema providers work on fields of schema encoded types.
+type translation struct {
+	C *Cyrillic
+	L *Latin
+	E *Ελληνικά
+}
+
+func ExampleRegisterSchemaProvider() {
```

Comment:
This function seems like it was meant to be implemented as a unit test, since this file ends in `_test.go`. If that's not the case I think it's reasonable to rename this file without the `_test` suffix.

Reply:
This is a whole file example, which appears in the Go Doc and is executed as part of the tests. It's required that the file is named _test.go so that it's not part of the main package compilation. 
See https://blog.golang.org/examples for details.

Reply:
Oooh, I didn't know about that, that's actually really cool.